core/state.go

```diff
@@ -437,3 +451,135 @@ func (s *State) ContractIsAlreadyDeployedAt(addr *felt.Felt, blockNumber uint64)
 	}
 	return deployedAt <= blockNumber, nil
 }
+
+func (s *State) Revert(blockNumber uint64, update *StateUpdate) error {
+	err := s.expectRoot(update.NewRoot)
+	if err != nil {
+		return err
+	}
+
+	// todo: remove classes from storage
+
+	// update declared classes trie
+	if err = s.updateDeclaredClasses(update.StateDiff.DeclaredV1Classes, true); err != nil {
+		return err
+	}
+
+	// update contracts
+	reversedDiff, err := s.buildReverseDiff(blockNumber, update.StateDiff)
+	if err != nil {
+		return err
+	}
+
+	if err = s.updateContracts(blockNumber, reversedDiff, false); err != nil {
+		return err
+	}
+
+	// purge deployed contracts
+	for _, contract := range update.StateDiff.DeployedContracts {
+		if err = s.purgeContract(contract.Address); err != nil {
+			return err
+		}
+	}
+
+	return s.expectRoot(update.OldRoot)
+}
+
+func (s *State) purgeContract(addr *felt.Felt) error {
+	contract, err := NewContract(addr, s.txn)
+	if err != nil {
+		return err
+	}
+
+	state, storageCloser, err := s.storage()
```

Comment:
is it ok if we don't call `storageCloser` in case of failure methods below?

Reply:
No it is not:) Will fix, thanks!